amazon/ionbenchmark/benchmark_spec.py

```diff
@@ -0,0 +1,241 @@
+       sample. Default: false.
+     * `protobuf_type` – the name of the protocol buffer type to use from the provided protocol buffer schema
+     * `descriptor_file` – A protocol buffer schema in the form of a `FileDescriptorSet` (generated by `protoc`
+       using the `--descriptor_set_out` option) that contains the necessary schema for reading/writing the protobuf
+       input file.
+     * `py_module` – A Python module path pointing to the module, generated by `protoc`, that contains the type
+       required to read/write the given input file.
+     * `py_file` – A file path pointing to the Python module, generated by `protoc`, that contains the type
+       required to read/write the given input file.
+
+    Note—for the `protobuf` format, the `protobuf_type` field and at least one of `descriptor_file`, `py_module`, or
+    `py_file` must be provided.
+    """
+    _data_object = None
+    _loader_dumper = None
+    _spec_working_directory = None
+
+    def __init__(self, params: dict, user_overrides: dict = None, user_defaults: dict = None, working_directory=None):
+        """
+        Construct a new BenchmarkSpec, possibly incorporating user supplied defaults or overrides.
+
+        Between the various dicts of parameters, the fields "format", "input_file", "command", "api", "iterations",
+        "warmups", and "io_type" must all have a value.
+
+        :param params: Values for this benchmark spec.
+        :param user_overrides: Values that override all other values.
+        :param user_defaults: Values that override the tool defaults, but not the `params`.
+        :param working_directory: reference point to use if `input_file` is a relative path. Defaults to os.getcwd().
+        """
+        if user_defaults is None:
+            user_defaults = {}
+        if user_overrides is None:
+            user_overrides = {}
+
+        self._spec_working_directory = working_directory or os.getcwd()
+
+        merged = { **_tool_defaults, **user_defaults, **params, **user_overrides }
+
+        # If not an absolute path, make relative to the working directory.
+        input_file = merged['input_file']
+        if not path.isabs(input_file):
+            input_file = path.join(self._spec_working_directory, input_file)
+            merged['input_file'] = input_file
+
+        # Convert symbols to strings
+        for k in merged.keys():
+            if isinstance(merged[k], IonPySymbol):
+                merged[k] = merged[k].text
+
+        super().__init__(merged)
+
+        for k in ["format", "input_file", "command", "api", "iterations", "warmups", "io_type"]:
+            if self[k] is None:
+                raise ValueError(f"Missing required parameter '{k}'")
+
+        if 'name' not in self:
+            self['name'] = f'({self.get_format()},{self.get_operation_name()},{path.basename(self.get_input_file())})'
+
+    def __missing__(self, key):
```

Comment:
fwiw i think you could inherit from `defaultdict` to do this. not sure it's any better.